lib/datadog/di/probe_manager.rb

```diff
@@ -0,0 +1,182 @@
+# frozen_string_literal: true
+
+module Datadog
+  module DI
+    # Stores probes received from remote config (that we can parse, in other
+    # wordrs, whose type/attributes we support), requests needed instrumentation
+    # for the probes via Instrumenter, and stores pending probes (those which
+    # haven't yet been instrumented successfully due to their targets not
+    # existing) and failed probes (where we are certain the target will not
+    # ever be loaded, or otherwise become valid).
+    #
+    # @api private
+    class ProbeManager
+      def initialize(settings, instrumenter, probe_notification_builder,
+        probe_notifier_worker, logger, telemetry: nil)
+        @settings = settings
+        @instrumenter = instrumenter
+        @probe_notification_builder = probe_notification_builder
+        @probe_notifier_worker = probe_notifier_worker
+        @logger = logger
+        @telemetry = telemetry
+        @installed_probes = {}
+        @pending_probes = {}
+        @failed_probes = {}
+
+        @definition_trace_point = TracePoint.trace(:end) do |tp|
+          begin
+            install_pending_method_probes(tp.self)
+          rescue => exc
+            raise if settings.dynamic_instrumentation.internal.propagate_all_exceptions
+            logger.warn("Unhandled exception in definition trace point: #{exc.class}: #{exc}")
+            telemetry&.report(exc, description: "Unhandled exception in definition trace point")
+            # TODO test this path
+          end
+        end
+      end
+
+      attr_reader :logger
+      attr_reader :telemetry
+
+      # TODO test that close is called during component teardown and
+      # the trace point is cleared
+      def close
+        definition_trace_point.disable
+        clear_hooks
+      end
+
+      def clear_hooks
+        pending_probes.clear
+        installed_probes.each do |probe_id, probe|
+          instrumenter.unhook(probe)
+        end
+        installed_probes.clear
+      end
+
+      attr_reader :settings
+      attr_reader :instrumenter
+      attr_reader :probe_notification_builder
+      attr_reader :probe_notifier_worker
+      attr_reader :installed_probes
+      attr_reader :pending_probes
+
+      # Probes that failed to instrument for reasons other than the target is
+      # not yet loaded are added to this collection, so that we do not try
+      # to instrument them every time remote configuration is processed.
+      attr_reader :failed_probes
+
+      # config is one probe info
```

Comment:
I think this comment is probably not accurate anymore to describe this function.

Reply:
Removed in https://github.com/DataDog/dd-trace-rb/pull/4097.